pgmpy/inference/Sampling.py

```diff
@@ -195,3 +197,192 @@ def likelihood_weighted_sample(self, evidence=None, size=1):
                     sampled[node] = list(map(lambda t: State(node, t),
                                          sample_discrete(states, cpd.values, size)))
         return sampled
+
+
+class GibbsSampling(MarkovChain):
+    """
+    Class for performing Gibbs sampling.
+
+    Parameters:
+    -----------
+    model: BayesianModel or MarkovModel
+        Model from which variables are inherited and transition probabilites computed.
+
+    Public Methods:
+    ---------------
+    set_start_state(state)
+    sample(start_state, size)
+    generate_sample(start_state, size)
+
+    Examples:
+    ---------
+    Initialization from a BayesianModel object:
+    >>> from pgmpy.factors import TabularCPD
+    >>> from pgmpy.models import BayesianModel
+    >>> intel_cpd = TabularCPD('intel', 2, [[0.7], [0.3]])
+    >>> sat_cpd = TabularCPD('sat', 2, [[0.95, 0.2], [0.05, 0.8]], evidence=['intel'], evidence_card=[2])
+    >>> student = BayesianModel()
+    >>> student.add_nodes_from(['intel', 'sat'])
+    >>> student.add_edge('intel', 'sat')
+    >>> student.add_cpds(intel_cpd, sat_cpd)
+
+    >>> from pgmpy.inference import GibbsSampling
+    >>> gibbs_chain = GibbsSampling(student)
+
+    Sample from it:
+    >>> gibbs_chain.sample(size=3)
+       intel  sat
+    0      0    0
+    1      0    0
+    2      1    1
+    """
+    def __init__(self, model=None):
+        super().__init__()
+        if isinstance(model, BayesianModel):
+            self._get_kernel_from_bayesian_model(model)
+        elif isinstance(model, MarkovModel):
+            self._get_kernel_from_markov_model(model)
+
+    def _get_kernel_from_bayesian_model(self, model):
+        """
+        Computes the Gibbs transition models from a Bayesian Network.
+        'Probabilistic Graphical Model Principles and Techniques', Koller and
+        Friedman, Section 12.3.3 pp 512-513.
+
+        Parameters:
+        -----------
+        model: BayesianModel
+            The model from which probabilities will be computed.
+        """
+        self.variables = np.array(model.nodes())
+        self.cardinalities = {var: model.get_cpds(var).variable_card for var in self.variables}
+        for var in self.variables:
+            other_vars = [v for v in self.variables if var != v]
+            other_cards = [range(self.cardinalities[v]) for v in self.variables if var != v]
+            cpds = [cpd for cpd in model.cpds if var in cpd.scope()]
+            prod_cpd = cpds[0]
+            for i in range(1, len(cpds)):
+                prod_cpd = prod_cpd.product(cpds[i])
+            kernel = {}
+            for tup in itertools.product(*other_cards):
+                scope = set(prod_cpd.scope())
```

Comment:
Put this line out of for loop. No need to compute it everytime since it's value remains same in all the loops.